GoogleForCreators/web-stories-wp
# Prepublish Text Contrast Check

## Summary
I added a text contrast check which checks the contrast/readability between the color palette provided by the part of the background image the text is sitting on and the text color. 

Here's it working: 

![ezgif-7-d5fa3b5f26af](https://user-images.githubusercontent.com/41136059/101434684-3fcdd200-38c8-11eb-86f7-bd71ea217840.gif)



## Relevant Technical Choices
- add an async calls section to `getPrepublishErrors` so I can use promises to put stuff on the DOM. Update the tests and implementations of `getPrepublishErrors` so it's async.
- Move the layout provider up to `app` so that prepublish can use the page size, important for getting data from the background images
- add helper functions and a text contrast check to `accessibility` in `prepublish` to do background image contrast checking. The helper functions create images based on text placement on the background, and also get the color palettes of the image to compare its luminosity to the font color. Then the colors are compared for contrast issues using the `contrastUtils`. They also help clean up the DOM after the calculations are made.
- refactored `getMediaBaseColor` module a bit to allow custom `onload` so I wasn't duplicating code in `accessibility`



## To-do

- [x] fix CORS issue, thanks Max and Pascal. 🙏
- [x] generalize for other layers (shapes, images, etc.) in between the text and bg image
- [x] generalize for different editor scales
- [x] refactor checklist run for async calls



## User-facing changes
When a user adds a background image to the page and places text on it, it should show a warning if the contrast is too low between the text and the image where the text sits. Same with shapes.



## Testing Instructions
1. Create a new story
2. Add a background image
3. Add text to the page
4. Place the text where it is hard to see, or change the color of the text to match the colors of the image
5. A warning should show in the prepublish checklist that the text contrast is too low
6. Repeat with a shape as a layer behind the text
7. Repeat with an image as a layer behind the text (it can be masked, rotated, flipped or not)



---



Fixes #5380
